### ISSUE TITLE
Rule proposal: prevent some DOM elements from having children or dangerouslySetInnerHTML

### ISSUE THREAD

**Issue by lencioni** [yannickcr/eslint-plugin-react#709]
There are some HTML elements that are normally self-closing (e.g. `img`, `br`, `hr`). If you try to give these children, React will give you a warning like

> Invariant Violation: img is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.

It would be nice to have a rule enforcing this.

Bad:

``` jsx
<img src="path/to/img.jpg">
  Children
</img>
```

``` jsx
<img src="path/to/img.jpg" dangerouslySetInnerHTML={{ __html: 'HTML' }} />
```

``` jsx
<img src="path/to/img.jpg" children="Children" />
```

Good:

``` jsx
<img src="path/to/img.jpg" />
```

The full list that React uses can be found:
- https://github.com/facebook/react/blob/85dcbf83/src/renderers/dom/shared/ReactDOMComponent.js#L421
- https://github.com/facebook/react/blob/85dcbf83/src/renderers/dom/shared/ReactDOMComponent.js#L449

Which comes out as:
- `area`
- `base`
- `br`
- `col`
- `embed`
- `hr`
- `img`
- `input`
- `keygen`
- `link`
- `menuitem`
- `meta`
- `param`
- `source`
- `track`
- `wbr`

**Comment by ljharb** [comment]
This seems like an excellent idea, and one that's mostly enforceable with both jsx and `React.createElement` (it wouldn't be enforceable with spreaded props that contain a "children" key, but I think that's OK)

**Comment by lencioni** [comment]
Yeah, I think the 99% solution is great.

**Comment by lencioni** [comment]
What should this be called? `void-element-tag-no-children`?

**Comment by ljharb** [comment]
Since it's just for DOM elements, what about `self-closing-dom-elements-no-children`? "void element" isn't a term I'm familiar with, but "self-closing" is.

**Comment by lencioni** [comment]
I thought about that, but I think self-closing may refer to elements that are also able to accept children but happen to be rendered without children, such as `div`. I'm not totally sold on the void naming either, but it is the language used by React and the HTML spec: https://www.w3.org/TR/html-markup/syntax.html#syntax-elements

**Comment by ljharb** [comment]
In that case, maybe `void-dom-elements-no-children`? "tag" isn't appropriate if it's not just jsx, and i like including "dom" to point out that it's not for custom components.

**Comment by lencioni** [comment]
That works for me!

**Comment by kepi0809** [comment]
why is `wbr` on this list??

**Comment by ljharb** [comment]
@kepi0809 because `<wbr />` is a self-closing element?

**Comment by kepi0809** [comment]
> @kepi0809 because `<wbr />` is a self-closing element?

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/wbr ?

**Comment by ljharb** [comment]
@kepi0809 HTML5 permits self-closing elements to omit the closing slash; anything is a self-closing element if it can not contain children. See "Tag omission" on that page.
